File: csharp/src/Vista.SDK/GmodVersioning.cs
using System.Diagnostics.CodeAnalysis;

namespace Vista.SDK;

internal sealed class GmodVersioning
{
    private readonly Dictionary<VisVersion, GmodVersioningNode> _versioningsMap = new();

    internal GmodVersioning(GmodVersioningDto dto)
    {
        foreach (var versioningDto in dto.Items)
        {
            var visVersion = versioningDto.Key;
            var gmodVersioningNode = new GmodVersioningNode(versioningDto.Value);
            _versioningsMap.Add(VisVersions.Parse(visVersion), gmodVersioningNode);
        }
    }

    public GmodNode? ConvertNode(VisVersion sourceVersion, GmodNode sourceNode, VisVersion targetVersion)
    {
        ValidateSourceAndTargetVersions(sourceVersion, targetVersion);

        GmodNode? node = sourceNode;
        var source = sourceVersion;

        while (source <= targetVersion - 1)
        {
            if (node is null)
                break;

            var target = source + 1;

            node = ConvertNodeInternal(source, node, target);
            source++;
        }

        return node;
    }

    public GmodNode? ConvertNodeInternal(VisVersion sourceVersion, GmodNode sourceNode, VisVersion targetVersion)
    {
        var nextCode = sourceNode.Code;

        if (TryGetVersioningNode(sourceVersion, out var versioningNode))
        {
            if (versioningNode.TryGetCodeChanges(sourceNode.Code, out var change))
            {
                if (targetVersion == change.NextVisVersion)
                    nextCode = change.NextCode ?? throw new InvalidOperationException("failed to set next code");
                else if (targetVersion == change.PreviousVisVersion)
                    nextCode =
                        change.PreviousCode ?? throw new InvalidOperationException("failed to set previous code");
            }
        }

        var targetGmod = VIS.Instance.GetGmod(targetVersion);

        if (!targetGmod.TryGetNode(nextCode, out var targetNode))
            return null;
        return targetNode.TryWithLocation(sourceNode.Location);
    }

    public LocalIdBuilder? ConvertLocalId(LocalIdBuilder sourceLocalId, VisVersion targetVersion)
    {
        if (sourceLocalId.VisVersion is null)
            throw new InvalidOperationException("Cant convert local ID without a specific VIS version");

        var targetLocalId = LocalIdBuilder.Create(targetVersion);

        if (sourceLocalId.PrimaryItem is not null)
        {
            var targetPrimaryitem = ConvertPath(
                sourceLocalId.VisVersion.Value,
                sourceLocalId.PrimaryItem,
                targetVersion
            );
            if (targetPrimaryitem is null)
                return null;
            targetLocalId = targetLocalId.WithPrimaryItem(targetPrimaryitem);
        }
        if (sourceLocalId.SecondaryItem is not null)
        {
            var targetSecondaryitem = ConvertPath(
                sourceLocalId.VisVersion.Value,
                sourceLocalId.SecondaryItem,
                targetVersion
            );
            if (targetSecondaryitem is null)
                return null;
            targetLocalId = targetLocalId.WithSecondaryItem(targetSecondaryitem);
        }

        return targetLocalId
            .WithVerboseMode(sourceLocalId.VerboseMode)
            .TryWithMetadataTag(sourceLocalId.Quantity)
            .TryWithMetadataTag(sourceLocalId.Content)
            .TryWithMetadataTag(sourceLocalId.Calculation)
            .TryWithMetadataTag(sourceLocalId.State)
            .TryWithMetadataTag(sourceLocalId.Command)
            .TryWithMetadataTag(sourceLocalId.Type)
            .TryWithMetadataTag(sourceLocalId.Position)
            .TryWithMetadataTag(sourceLocalId.Detail);
    }

    public LocalId? ConvertLocalId(LocalId sourceLocalId, VisVersion targetVersion) =>
        ConvertLocalId(sourceLocalId.Builder, targetVersion)?.Build();

    public GmodPath? ConvertPath(VisVersion sourceVersion, GmodPath sourcePath, VisVersion targetVersion)
    {
        var targetEndNode = ConvertNode(sourceVersion, sourcePath.Node, targetVersion);
        if (targetEndNode is null)
            return null;

        if (targetEndNode.IsRoot)
            return new GmodPath(targetEndNode._parents, targetEndNode, skipVerify: true);

        var targetGmod = VIS.Instance.GetGmod(targetVersion);
        var sourceGmod = VIS.Instance.GetGmod(sourceVersion);

        var qualifyingNodes = sourcePath
            .GetFullPath()
            .Select((t, i) => (SourceNode: t.Node, TargetNode: ConvertNode(sourceVersion, t.Node, targetVersion)!))
            .ToArray();
        if (qualifyingNodes.Any(t => t.TargetNode is null))
            throw new Exception("Could convert node forward");

        var potentialParents = qualifyingNodes.Select(n => n.TargetNode).Take(qualifyingNodes.Length - 1).ToList();
        if (GmodPath.IsValid(potentialParents, targetEndNode))
            return new GmodPath(potentialParents, targetEndNode, skipVerify: true);

        var path = new List<GmodNode>();
        for (int i = 0; i <= qualifyingNodes.Length - 1; i++)
        {
            var qualifyingNode = qualifyingNodes[i];

            if (i > 0 && qualifyingNode.TargetNode.Code == qualifyingNodes[i - 1].TargetNode.Code)
                continue;

            var codeChanged = qualifyingNode.SourceNode.Code != qualifyingNode.TargetNode.Code;

            var sourceNormalAssignment = qualifyingNode.SourceNode.ProductType;
            var targetNormalAssignment = qualifyingNode.TargetNode.ProductType;

            var normalAssignmentChanged = sourceNormalAssignment?.Code != targetNormalAssignment?.Code;

            var selectionChanged = false;

            static void AddToPath(Gmod targetGmod, List<GmodNode> path, GmodNode node)
            {
                if (path.Count > 0)
                {
                    var prev = path[path.Count - 1];
                    if (!prev.IsChild(node))
                    {
                        for (int j = path.Count - 1; j >= 0; j--)
                        {
                            var parent = path[j];
                            var currentParents = path.Take(j + 1).ToArray();
                            if (!targetGmod.PathExistsBetween(currentParents, node, out var remaining))
                            {
                                if (!currentParents.Any(n => n.IsAssetFunctionNode && n.Code != parent.Code))
                                    throw new Exception("Tried to remove last asset function node");
                                path.RemoveAt(j);
                            }
                            else
                            {
                                path.AddRange(remaining);
                                break;
                            }
                        }
                    }
                }

                path.Add(node);
            }

            if (codeChanged)
            {
                AddToPath(targetGmod, path, qualifyingNode.TargetNode);
            }
            else if (normalAssignmentChanged) // AC || AN || AD
            {
                var wasDeleted = sourceNormalAssignment is not null && targetNormalAssignment is null;

                if (!codeChanged)
                    AddToPath(targetGmod, path, qualifyingNode.TargetNode);

                if (wasDeleted)
                {
                    if (qualifyingNode.TargetNode.Code == targetEndNode.Code)
                    {
                        var next = qualifyingNodes[i + 1];
                        if (next.TargetNode.Code != qualifyingNode.TargetNode.Code)
                            throw new Exception("Normal assignment end node was deleted");
                    }
                    i++;
                }
                else if (qualifyingNode.TargetNode.Code != targetEndNode.Code)
                {
                    AddToPath(targetGmod, path, targetNormalAssignment!);
                    i++; // Holy moly
                }
            }
            if (selectionChanged) // SC || SN || SD
            { }

            if (!codeChanged && !normalAssignmentChanged)
            {
                AddToPath(targetGmod, path, qualifyingNode.TargetNode);
            }

            if (path[path.Count - 1].Code == targetEndNode.Code)
                break;
        }

        potentialParents = path.Take(path.Count - 1).ToList();
        targetEndNode = path.Last();

        if (!GmodPath.IsValid(potentialParents, targetEndNode, out var missinkLinkAt))
            throw new Exception($"Didnt end up with valid path for {sourcePath}");

        return new GmodPath(potentialParents, targetEndNode);
    }

    private bool HasChangesInNextVisVersion(VisVersion source)
    {
        var target = source + 1;

        return TryGetVersioningNode(target, out _);
    }

    private bool TryGetVersioningNode(
        VisVersion visVersion,
        [MaybeNullWhen(false)] out GmodVersioningNode versioningNode
    ) => _versioningsMap.TryGetValue(visVersion, out versioningNode);

    private void ValidateSourceAndTargetVersions(VisVersion sourceVersion, VisVersion targetVersion)
    {
        if (string.IsNullOrWhiteSpace(sourceVersion.ToVersionString()))
            throw new ArgumentException("Invalid source VIS Version: " + sourceVersion.ToVersionString());

        if (string.IsNullOrWhiteSpace(targetVersion.ToVersionString()))
            throw new ArgumentException("Invalid target VISVersion: " + targetVersion.ToVersionString());
    }

    private readonly record struct GmodVersioningNode
    {
        private readonly Dictionary<string, GmodVersioningNodeChanges> _versioningNodeChanges = new();

        internal GmodVersioningNode(IReadOnlyDictionary<string, GmodVersioningNodeChangesDto> dto)
        {
            foreach (var versioningNodeDto in dto)
            {
                if (
                    versioningNodeDto.Value is null
                    || (versioningNodeDto.Value.PreviousCode is null && versioningNodeDto.Value.NextCode is null)
                )
                    continue;

                var code = versioningNodeDto.Key;
                var versioningNodeChanges = new GmodVersioningNodeChanges(
                    VisVersions.TryParse(versioningNodeDto.Value.NextVisVersion, out var nextVisVersion)
                        ? nextVisVersion
                        : null,
                    versioningNodeDto.Value.NextCode,
                    VisVersions.TryParse(versioningNodeDto.Value.PreviousVisVersion, out var previoiusVisVersion)
                        ? previoiusVisVersion
                        : null,
                    versioningNodeDto.Value.PreviousCode
                );
                _versioningNodeChanges.Add(code, versioningNodeChanges);
            }
        }

        public bool TryGetCodeChanges(string code, [MaybeNullWhen(false)] out GmodVersioningNodeChanges nodeChanges) =>
            _versioningNodeChanges.TryGetValue(code, out nodeChanges);
    }

    private sealed record GmodVersioningNodeChanges(
        VisVersion? NextVisVersion,
        string? NextCode,
        VisVersion? PreviousVisVersion,
        string? PreviousCode
    );
}

File: cpp/src/dnv/vista/sdk/GmodVersioning.cpp
/**
 * @file GmodVersioning.cpp
 * @brief Implementation of the GmodVersioning class for converting GMOD objects between VIS versions.
 */

#include "pch.h"

#include "dnv/vista/sdk/GmodVersioning.h"
#include "dnv/vista/sdk/internal/PathConversionCache.h"

#include "dnv/vista/sdk/Gmod.h"
#include "dnv/vista/sdk/GmodNode.h"
#include "dnv/vista/sdk/GmodPath.h"
#include "dnv/vista/sdk/LocalId.h"
#include "dnv/vista/sdk/LocalIdBuilder.h"
#include "dnv/vista/sdk/MetadataTag.h"
#include "dnv/vista/sdk/VIS.h"
#include "dnv/vista/sdk/VISVersion.h"

namespace dnv::vista::sdk
{
	//=====================================================================
	// Static helper functions
	//=====================================================================

	thread_local static std::vector<const GmodNode*> t_currentParentsBuffer;
	thread_local static std::vector<const GmodNode*> t_remainingBuffer;

	static void addToPath( const Gmod& gmod, std::vector<GmodNode>& path, const GmodNode& node )
	{
		if ( !path.empty() )
		{
			const GmodNode& prev = path.back();
			if ( !prev.isChild( node ) )
			{
				for ( int j = static_cast<int>( path.size() ) - 1; j >= 0; --j )
				{
					const GmodNode& parent = path[static_cast<size_t>( j )];

					t_currentParentsBuffer.clear();
					const size_t currentParentsCount = static_cast<size_t>( j + 1 );
					t_currentParentsBuffer.reserve( currentParentsCount );

					for ( size_t k = 0; k < currentParentsCount; ++k )
					{
						t_currentParentsBuffer.push_back( &path[k] );
					}

					t_remainingBuffer.clear();
					t_remainingBuffer.reserve( 16 );

					if ( !gmod.pathExistsBetween( t_currentParentsBuffer, node, t_remainingBuffer ) )
					{
						bool hasOtherAssetFunction = false;
						const std::string_view parentCode = parent.code();
						for ( const GmodNode* pathNode : t_currentParentsBuffer )
						{
							if ( pathNode->isAssetFunctionNode() && pathNode->code() != parentCode )
							{
								hasOtherAssetFunction = true;
								break;
							}
						}

						if ( !hasOtherAssetFunction )
						{
							throw std::runtime_error( "Tried to remove last asset function node" );
						}
						path.erase( path.begin() + static_cast<std::ptrdiff_t>( j ) );
					}
					else
					{
						const auto nodeLocation = node.location();
						if ( nodeLocation.has_value() )
						{
							path.reserve( path.size() + t_remainingBuffer.size() );

							for ( const GmodNode* n : t_remainingBuffer )
							{
								if ( n->isIndividualizable( false, true ) )
								{
									path.emplace_back( std::move( n->tryWithLocation( nodeLocation.value() ) ) );
								}
								else
								{
									path.emplace_back( *n );
								}
							}
						}
						else
						{
							path.reserve( path.size() + t_remainingBuffer.size() );
							for ( const GmodNode* n : t_remainingBuffer )
							{
								path.emplace_back( *n );
							}
						}
						break;
					}
				}
			}
		}

		path.emplace_back( node );
	}

	//=====================================================================
	// GmodVersioning class
	//=====================================================================

	//----------------------------------------------
	// Construction
	//----------------------------------------------

	GmodVersioning::GmodVersioning( const utils::StringMap<GmodVersioningDto>& dto )
	{
		m_versioningsMap.reserve( dto.size() );

		for ( const auto& [versionStr, versioningDto] : dto )
		{
			VisVersion version = VisVersionExtensions::parse( versionStr );

			m_versioningsMap.insertOrAssign( version, GmodVersioningNode( version, versioningDto.items() ) );
		}
	}

	//----------------------------------------------
	// Conversion
	//----------------------------------------------

	//----------------------------
	// Node
	//----------------------------

	std::optional<GmodNode> GmodVersioning::convertNode(
		VisVersion sourceVersion, const GmodNode& sourceNode, VisVersion targetVersion ) const
	{
		if ( sourceNode.code().empty() )
		{
			return std::nullopt;
		}

		validateSourceAndTargetVersions( sourceVersion, targetVersion );

		std::optional<GmodNode> node = sourceNode;
		VisVersion source = sourceVersion;

		while ( source <= targetVersion - 1 )
		{
			if ( !node.has_value() )
			{
				break;
			}

			const VisVersion target = source + 1;

			node = convertNodeInternal( source, *node, target );

			++source;
		}

		return node;
	}

	std::optional<GmodNode> GmodVersioning::convertNode(
		VisVersion sourceVersion, const GmodNode& sourceNode, VisVersion targetVersion, const Gmod& targetGmod ) const
	{
		if ( sourceNode.code().empty() )
		{
			return std::nullopt;
		}

		validateSourceAndTargetVersions( sourceVersion, targetVersion );

		std::optional<GmodNode> node = sourceNode;
		VisVersion source = sourceVersion;

		while ( source <= targetVersion - 1 )
		{
			if ( !node.has_value() )
			{
				break;
			}

			const VisVersion target = source + 1;

			if ( target == targetVersion )
			{
				node = convertNodeInternal( source, *node, target, targetGmod );
			}
			else
			{
				node = convertNodeInternal( source, *node, target );
			}

			++source;
		}

		return node;
	}

	//----------------------------
	// Path
	//----------------------------

	std::optional<GmodPath> GmodVersioning::convertPath( VisVersion sourceVersion, const GmodPath& sourcePath, VisVersion targetVersion ) const
	{
		auto& pathCache = internal::PathConversionCache::instance();
		const std::string pathString = sourcePath.toString();

		std::optional<GmodPath> cachedResult;
		if ( pathCache.tryGetCachedConversion( sourceVersion, pathString, targetVersion, cachedResult ) )
		{
			return cachedResult;
		}

		auto result = convertPathInternal( sourceVersion, sourcePath, targetVersion );

		pathCache.cacheConversion( sourceVersion, pathString, targetVersion, result );

		return result;
	}

	std::optional<GmodPath> GmodVersioning::convertPathInternal( VisVersion sourceVersion, const GmodPath& sourcePath, VisVersion targetVersion ) const
	{
		std::optional<GmodNode> targetEndNode = convertNode( sourceVersion, sourcePath.node(), targetVersion );
		if ( !targetEndNode.has_value() )
		{
			return std::nullopt;
		}

		auto& vis = VIS::instance();
		const auto& targetGmod = vis.gmod( targetVersion );

		if ( targetEndNode->isRoot() )
		{
			const GmodNode* rootNodeInGmodPtr = nullptr;
			if ( !targetGmod.tryGetNode( targetEndNode->code(), rootNodeInGmodPtr ) || !rootNodeInGmodPtr )
			{
				throw std::runtime_error( "Failed to get root node from targetGmod during root path conversion" );
			}

			return GmodPath( targetGmod, *rootNodeInGmodPtr, {} );
		}

		std::vector<std::pair<const GmodNode*, GmodNode>> qualifyingNodes;

		auto fullPathEnumerator = sourcePath.fullPath();
		qualifyingNodes.reserve( 16 );

		while ( fullPathEnumerator.next() )
		{
			const auto& [depth, nodePtr] = fullPathEnumerator.current();
			if ( !nodePtr )
			{
				continue;
			}

			std::optional<GmodNode> convertedNodeOpt = convertNode( sourceVersion, *nodePtr, targetVersion, targetGmod );
			if ( !convertedNodeOpt.has_value() )
			{
				throw std::runtime_error( "Could not convert node forward" );
			}

			qualifyingNodes.emplace_back( nodePtr, std::move( *convertedNodeOpt ) );
		}

		std::vector<GmodNode> potentialParents;
		potentialParents.reserve( qualifyingNodes.size() > 0 ? qualifyingNodes.size() - 1 : 0 );

		for ( size_t i = 0; i < qualifyingNodes.size() - 1; ++i )
		{
			potentialParents.emplace_back( qualifyingNodes[i].second );
		}

		std::vector<GmodNode*> potentialParentPtrs;
		potentialParentPtrs.reserve( potentialParents.size() );
		for ( auto& parent : potentialParents )
		{
			potentialParentPtrs.push_back( &parent );
		}

		if ( GmodPath::isValid( potentialParentPtrs, *targetEndNode ) )
		{
			return GmodPath( targetGmod, *targetEndNode, std::move( potentialParents ) );
		}

		std::vector<GmodNode> path;
		path.reserve( 64 );

		for ( size_t i = 0; i < qualifyingNodes.size(); ++i )
		{
			const auto& qualifyingNode = qualifyingNodes[i];

			if ( i > 0 && qualifyingNode.second.code() == qualifyingNodes[i - 1].second.code() )
			{
				continue;
			}

			const bool codeChanged = qualifyingNode.first->code() != qualifyingNode.second.code();

			const std::optional<GmodNode> sourceNormalAssignment = qualifyingNode.first->productType();
			const std::optional<GmodNode> targetNormalAssignment = qualifyingNode.second.productType();

			const bool normalAssignmentChanged =
				sourceNormalAssignment.has_value() != targetNormalAssignment.has_value() ||
				( sourceNormalAssignment.has_value() && targetNormalAssignment.has_value() &&
					sourceNormalAssignment->code() != targetNormalAssignment->code() );

			bool selectionChanged = false;

			if ( codeChanged )
			{
				addToPath( targetGmod, path, qualifyingNode.second );
			}
			else if ( normalAssignmentChanged )
			{
				/* SC || SN || SD */

				bool wasDeleted = sourceNormalAssignment.has_value() && !targetNormalAssignment.has_value();

				if ( !codeChanged )
				{
					addToPath( targetGmod, path, qualifyingNode.second );
				}

				if ( wasDeleted )
				{
					if ( qualifyingNode.second.code() == targetEndNode->code() )
					{
						if ( i + 1 < qualifyingNodes.size() )
						{
							const auto& next = qualifyingNodes[i + 1];
							if ( next.second.code() != qualifyingNode.second.code() )
							{
								throw std::runtime_error( "Normal assignment end node was deleted" );
							}
						}
					}
					continue;
				}
				else if ( qualifyingNode.second.code() != targetEndNode->code() )
				{
					if ( targetNormalAssignment.has_value() )
					{
						GmodNode targetNormalAssignmentVal = *targetNormalAssignment;
						if ( qualifyingNode.second.location().has_value() && targetNormalAssignmentVal.isIndividualizable( false, true ) )
						{
							targetNormalAssignmentVal = targetNormalAssignmentVal.tryWithLocation( *qualifyingNode.second.location() );
						}
						addToPath( targetGmod, path, targetNormalAssignmentVal );

						++i;
					}
				}
			}

			if ( selectionChanged )
			{
				/* SC || SN || SD */
			}

			if ( !codeChanged && !normalAssignmentChanged )
			{
				addToPath( targetGmod, path, qualifyingNode.second );
			}

			if ( !path.empty() && path.back().code() == targetEndNode->code() )
			{
				break;
			}
		}

		if ( path.empty() || path.size() < 1 )
		{
			throw std::runtime_error( "Path reconstruction resulted in an empty path" );
		}

		if ( path.size() == 1 )
		{
			return GmodPath( targetGmod, std::move( path[0] ), {} );
		}

		std::vector<GmodNode> potentialParentsFromPath;
		potentialParentsFromPath.reserve( path.size() - 1 );

		for ( size_t i = 0; i < path.size() - 1; ++i )
		{
			potentialParentsFromPath.emplace_back( std::move( path[i] ) );
		}

		GmodNode targetEndNodeFromPath = std::move( path.back() );

		std::vector<GmodNode*> potentialParentPtrsFromPath;
		potentialParentPtrsFromPath.reserve( potentialParentsFromPath.size() );
		for ( auto& parent : potentialParentsFromPath )
		{
			potentialParentPtrsFromPath.push_back( &parent );
		}

		int missingLinkAt;
		if ( !GmodPath::isValid( potentialParentPtrsFromPath, targetEndNodeFromPath, missingLinkAt ) )
		{
			throw std::runtime_error( "Did not end up with a valid path" );
		}

		return GmodPath( targetGmod, std::move( targetEndNodeFromPath ), std::move( potentialParentsFromPath ) );
	}

	//----------------------------
	// Local Id
	//----------------------------

	std::optional<LocalIdBuilder> GmodVersioning::convertLocalId(
		const LocalIdBuilder& sourceLocalId, VisVersion targetVersion ) const
	{
		if ( !sourceLocalId.visVersion().has_value() )
		{
			throw std::invalid_argument( "Cannot convert local ID without a specific VIS version" );
		}

		LocalIdBuilder targetLocalId = LocalIdBuilder::create( targetVersion );

		if ( sourceLocalId.primaryItem().has_value() )
		{
			auto targetPrimaryItem = convertPath(
				*sourceLocalId.visVersion(),
				sourceLocalId.primaryItem().value(),
				targetVersion );

			if ( !targetPrimaryItem.has_value() )
			{
				return std::nullopt;
			}

			targetLocalId = targetLocalId.withPrimaryItem( std::move( *targetPrimaryItem ) );
		}

		if ( sourceLocalId.secondaryItem().has_value() )
		{
			auto targetSecondaryItem = convertPath(
				*sourceLocalId.visVersion(),
				sourceLocalId.secondaryItem().value(),
				targetVersion );

			if ( !targetSecondaryItem.has_value() )
			{
				return std::nullopt;
			}

			targetLocalId = targetLocalId.withSecondaryItem( std::move( *targetSecondaryItem ) );
		}

		return targetLocalId
			.withVerboseMode( sourceLocalId.isVerboseMode() )
			.tryWithMetadataTag( sourceLocalId.quantity() )
			.tryWithMetadataTag( sourceLocalId.content() )
			.tryWithMetadataTag( sourceLocalId.calculation() )
			.tryWithMetadataTag( sourceLocalId.state() )
			.tryWithMetadataTag( sourceLocalId.command() )
			.tryWithMetadataTag( sourceLocalId.type() )
			.tryWithMetadataTag( sourceLocalId.position() )
			.tryWithMetadataTag( sourceLocalId.detail() );
	}

	std::optional<LocalId> GmodVersioning::convertLocalId(
		const LocalId& sourceLocalId, VisVersion targetVersion ) const
	{
		auto builder = convertLocalId( sourceLocalId.builder(), targetVersion );

		return builder.has_value()
				   ? std::make_optional( builder->build() )
				   : std::nullopt;
	}

	//----------------------------------------------
	// GmodVersioning::GmodVersioningNode class
	//----------------------------------------------

	//----------------------------
	// Construction
	//----------------------------

	GmodVersioning::GmodVersioningNode::GmodVersioningNode(
		VisVersion visVersion,
		const utils::StringMap<GmodNodeConversionDto>& dto )
		: m_visVersion{ visVersion }
	{
		for ( const auto& [code, dtoNode] : dto )
		{
			GmodNodeConversion conversion;
			conversion.source = dtoNode.source();
			if ( dtoNode.target().empty() )
			{
				conversion.target = std::nullopt;
			}
			else
			{
				conversion.target = dtoNode.target();
			}
			conversion.oldAssignment = dtoNode.oldAssignment();
			conversion.newAssignment = dtoNode.newAssignment();
			conversion.deleteAssignment = dtoNode.deleteAssignment();

			if ( !dtoNode.operations().empty() )
			{
				for ( const auto& type : dtoNode.operations() )
				{
					conversion.operations.insert( parseConversionType( type ) );
				}
			}

			m_versioningNodeChanges.emplace( code, conversion );
		}
	}

	//----------------------------
	// Private helper methods
	//----------------------------

	std::optional<GmodNode> GmodVersioning::convertNodeInternal(
		[[maybe_unused]] VisVersion sourceVersion, const GmodNode& sourceNode, VisVersion targetVersion ) const
	{
		validateSourceAndTargetVersionPair( sourceNode.visVersion(), targetVersion );

		std::string_view nextCodeView = sourceNode.code();

		const GmodVersioningNode* versioningNode = nullptr;
		if ( tryGetVersioningNode( targetVersion, versioningNode ) )
		{
			const GmodNodeConversion* change = nullptr;
			if ( versioningNode->tryGetCodeChanges( nextCodeView, change ) && change && change->target.has_value() )
			{
				nextCodeView = change->target.value();
			}
		}

		const auto& targetGmod = VIS::instance().gmod( targetVersion );

		const GmodNode* targetNodePtr = nullptr;
		if ( !targetGmod.tryGetNode( nextCodeView, targetNodePtr ) )
		{
			return std::nullopt;
		}

		if ( sourceNode.location().has_value() )
		{
			return targetNodePtr->tryWithLocation( sourceNode.location().value() );
		}

		return *targetNodePtr;
	}

	std::optional<GmodNode> GmodVersioning::convertNodeInternal(
		[[maybe_unused]] VisVersion sourceVersion, const GmodNode& sourceNode, VisVersion targetVersion, const Gmod& targetGmod ) const
	{
		validateSourceAndTargetVersionPair( sourceNode.visVersion(), targetVersion );

		std::string_view nextCodeView = sourceNode.code();

		const GmodVersioningNode* versioningNode = nullptr;
		if ( tryGetVersioningNode( targetVersion, versioningNode ) )
		{
			const GmodNodeConversion* change = nullptr;
			if ( versioningNode->tryGetCodeChanges( nextCodeView, change ) && change && change->target.has_value() )
			{
				nextCodeView = change->target.value();
			}
		}

		const GmodNode* targetNodePtr = nullptr;
		if ( !targetGmod.tryGetNode( nextCodeView, targetNodePtr ) )
		{
			return std::nullopt;
		}

		if ( sourceNode.location().has_value() )
		{
			return targetNodePtr->tryWithLocation( sourceNode.location().value() );
		}

		return *targetNodePtr;
	}
}
